multisig-wallet/src/main.sw

```diff
@@ -147,11 +147,51 @@ impl MultiSignatureWallet for Contract {
         true
     }
 
-    /// Takes in transaction data and hashes it into a unique tx hash
-    /// Used for verification of message
-    fn get_transaction_hash(to: ContractId, value: u64, data: b256, nonce: u64) -> b256 {
-        // TODO: data > b256?
-        _get_transaction_hash(to, value, data, nonce, contract_id())
+    /// Transfers assets to outputs & contracts
+    ///
+    /// # Panics
+    ///
+    /// - When the constructor has not been called to initialize the contract
+    /// - When the balance of the asset being sent is less than the balance in the contract
+    /// - When the public key cannot be recovered from a signature
+    /// - When the recovered addresses are not in ascending order (0x1 < 0x2 < 0x3...)
+    /// - When the signer is not an owner
+    /// - When the total approval count is less than the required threshold for execution
+    fn transfer(to: Sender, asset_id: ContractId, value: u64, data: b256, signature1: B512, signature2: B512) -> bool {
+        require(storage.nonce != 0, Error::NotInitialized);
+        require(value <= this_balance(asset_id), Error::InsufficientAssetAmount);
+
+        let tx_hash = _get_transaction_hash(to, value, data, storage.nonce, contract_id());
+
+        // TODO: change to Vec<B512> once implemented and then iterate instead of hardcoding length
+        let signer1: b256 = match ec_recover_address(signature1, tx_hash) {
+            Result::Ok(address) => address.value, _ => revert(42), 
+        };
+
+        let signer2: b256 = match ec_recover_address(signature2, tx_hash) {
+            Result::Ok(address) => address.value, _ => revert(42), 
+        };
+
+        require(~b256::min() < signer1 && signer1 < signer2, Error::IncorrectSignerOrdering);
```

Comment:
Why is it required that the recovered address of signer1 is less than signer2?  Also could we implement some check here say:
```
if signer1 < signer2 {
    swap(signer1, signer2)
}
```

Reply:
Otherwise you can submit your own signature multiple times to game the system and send any Tx through that you want. Making sure that they are increasing in size means that they cannot be 0 and cannot be the same address and thus likely to come from different owners... or at least the same owner but with different addresses